Svelte keeps port 5000 on OSX while killed

Sometimes Svelte in `npm run dev` mode crashes. This happens for example when doing a NPM install. This is not a huge problem as you can quickly restart. But then it takes another port which is frustrating:

**Logs**

```bash
Your application is ready~! 🚀

  ➡ Port 5000 is taken; using 61945 instead

```

**To Reproduce**

Let Svelte crash. There is not specific code to do this. It mostly happens when changing NPM or rollup configs.

**To Fix**

```bash
kill $(lsof -ti:5000) 
```

Best thing would be if Svelte would be killing it's orphaned process itself. Now we do that by hand with the command above this line.

**Expected behavior**
Start at port 5000

**Stacktraces**

This is the error on crash:

```bash
[routify] Watching /Users/xxxxxxxxx/Software/client/xxxxxxxxx-client/src/pages
Error: listen EADDRINUSE: address already in use :::35729
at Server.setupListenHandle [as _listen2] (net.js:1306:16)
at listenInCluster (net.js:1354:12)
at Server.listen (net.js:1442:7)
at Server.listen (/Users/xxxxxxxxx/Software/client/xxxxxxxxx-client/node_modules/livereload/lib/livereload.js:80:28)
at Object.exports.createServer (/Users/xxxxxxxxx/Software/client/xxxxxxxxx-client/node_modules/livereload/lib/livereload.js:245:14)
at livereload (/Users/xxxxxxxxx/Software/client/xxxxxxxxx-client/node_modules/rollup-plugin-livereload/dist/index.cjs.js:20:29)
at Object.<anonymous> (/Users/xxxxxxxxx/Software/client/xxxxxxxxx-client/rollup.config.js:141:18)
at Module._compile (internal/modules/cjs/loader.js:955:30)
at Object.require.extensions.<computed> (/Users/xxxxxxxxx/Software/client/xxxxxxxxx-client/node_modules/rollup/dist/bin/rollup:824:24)
at Module.load (internal/modules/cjs/loader.js:811:32) {
  code: 'EADDRINUSE',
    errno: 'EADDRINUSE',
    syscall: 'listen',
    address: '::',
    port: 35729
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! myimin@1.0.0 dev: `rollup -c -w`
npm ERR! Exit status 1
npm ERR!
  npm ERR! Failed at the myimin@1.0.0 dev script.
  npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

  npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/xxxxxxxxx/.npm/_logs/2020-04-06T14_01_07_666Z-debug.log

```

It seems to be triggered by the Routify watcher.

The mentioned log file:

```bash
0 info it worked if it ends with ok
  1 verbose cli [ '/usr/local/bin/node', '/usr/local/bin/npm', 'run', 'dev' ]
2 info using npm@6.13.4
3 info using node@v12.14.1
4 verbose run-script [ 'predev', 'dev', 'postdev' ]
5 info lifecycle xxxxxxxxx@1.0.0~predev: xxxxxxxxx@1.0.0
6 info lifecycle xxxxxxxxx@1.0.0~dev: xxxxxxxxx@1.0.0
7 verbose lifecycle xxxxxxxxx@1.0.0~dev: unsafe-perm in lifecycle true
8 verbose lifecycle xxxxxxxxx@1.0.0~dev: PATH: /usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/Users/xxxxxxxxx/Software/client/xxxxxxxxx-client/node_modules/.bin:/Users/xxxxxxxxx/.yarn/bin:/Users/xxxxxxxxx/.config/yarn/global/node_modules/.bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
9 verbose lifecycle xxxxxxxxx@1.0.0~dev: CWD: /Users/xxxxxxxxx/Software/client/xxxxxxxxx-client
10 silly lifecycle xxxxxxxxx@1.0.0~dev: Args: [ '-c', 'rollup -c -w' ]
11 silly lifecycle xxxxxxxxx@1.0.0~dev: Returned: code: 1  signal: null
12 info lifecycle xxxxxxxxx@1.0.0~dev: Failed to exec dev script
13 verbose stack Error: xxxxxxxxx@1.0.0 dev: `rollup -c -w`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:332:16)
13 verbose stack     at EventEmitter.emit (events.js:223:5)
13 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:223:5)
13 verbose stack     at maybeClose (internal/child_process.js:1021:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
14 verbose pkgid xxxxxxxxx@1.0.0
15 verbose cwd /Users/xxxxxxxxx/Software/client/xxxxxxxxx-client
16 verbose Darwin 19.4.0
17 verbose argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "dev"
18 verbose node v12.14.1
19 verbose npm  v6.13.4
20 error code ELIFECYCLE
21 error errno 1
22 error xxxxxxxxx@1.0.0 dev: `rollup -c -w`
22 error Exit status 1
23 error Failed at the xxxxxxxxx@1.0.0 dev script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]

```

**Information about your Svelte project:**
- Your operating system:OS X 10.15.4

- Svelte version3.20.1

- Rollup 

**Severity**

Annoying while developing.

--- a/rollup.config.js
+++ b/rollup.config.js
@@ -3,8 +3,29 @@ import resolve from '@rollup/plugin-node-resolve';
 import commonjs from '@rollup/plugin-commonjs';
 import livereload from 'rollup-plugin-livereload';
 import { terser } from 'rollup-plugin-terser';
+import { spawn } from 'child_process';
 
 const production = !process.env.ROLLUP_WATCH;
+
+function serve() {
+	let server;
+
+	return {
+		buildEnd(error) {
+			if (error && server) {
+				server.kill();
+			}
+		},
+		writeBundle() {
+			if (!server) {
+				server = spawn('npm', ['run', 'start', '--', '-d'], {
+					stdio: ['ignore', 'inherit', 'inherit'],
+					shell: true
+				});
+			}
+		}
+	}
+}
 
 export default {
 	input: 'src/main.js',
@@ -52,20 +73,3 @@ export default {
 		clearScreen: false
 	}
 };
-
-function serve() {
-	let started = false;
-
-	return {
-		writeBundle() {
-			if (!started) {
-				started = true;
-
-				require('child_process').spawn('npm', ['run', 'start', '--', '--dev'], {
-					stdio: ['ignore', 'inherit', 'inherit'],
-					shell: true
-				});
-			}
-		}
-	};
-}

--- a/rollup.config.js
+++ b/rollup.config.js
@@ -3,28 +3,28 @@ import resolve from '@rollup/plugin-node-resolve';
 import commonjs from '@rollup/plugin-commonjs';
 import livereload from 'rollup-plugin-livereload';
 import { terser } from 'rollup-plugin-terser';
-import { spawn } from 'child_process';
 
 const production = !process.env.ROLLUP_WATCH;
 
 function serve() {
 	let server;
+	
+	function toExit() {
+		if (server) server.kill(0);
+	}
 
 	return {
-		buildEnd(error) {
-			if (error && server) {
-				server.kill();
-			}
-		},
 		writeBundle() {
-			if (!server) {
-				server = spawn('npm', ['run', 'start', '--', '-d'], {
-					stdio: ['ignore', 'inherit', 'inherit'],
-					shell: true
-				});
-			}
+			if (server) return;
+			server = require('child_process').spawn('npm', ['run', 'start', '--', '--dev'], {
+				stdio: ['ignore', 'inherit', 'inherit'],
+				shell: true
+			});
+
+			process.on('SIGTERM', toExit);
+			process.on('exit', toExit);
 		}
-	}
+	};
 }
 
 export default {

